File: features/step_definitions/job_configuration_steps.rb

When /^I configure the job$/ do
  @job.configure
end

When /^I add a shell build step "([^"]*)"$/ do |script|
  @step = @job.add_shell_step(script)
end

When /^I add a shell build step$/ do |script|
  @step = @job.add_shell_step(script)
end

When /^I add a shell build step "([^"]*)" in the job configuration$/ do |script|
  @job.configure do
    @step = @job.add_shell_step(script)
  end
end

When /^I add a shell build step in the job configuration$/ do |script|
  @job.configure do
    @step = @job.add_shell_step(script)
  end
end

When /^I change a shell build step to "([^"]*)"$/ do |script|
  @step.command script
end


When /^I add always fail build step$/ do
  @step = @job.add_shell_step "exit 1"
end

When /^I add always fail build step in the job configuration$/ do
  @job.configure do
    step 'I add always fail build step'
  end
end

When /^I add "([^"]*)" build action$/ do |action|
  @job.add_build_action(action)
end

When /^I add "([^"]*)" post-build action$/ do |action|
  @job.add_postbuild_action(action)
end

When /^I tie the job to the "([^"]*)" label$/ do |label|
  @job.configure do
    @job.label_expression = label
  end
end

When /^I tie the job to the slave$/ do
  step %{I tie the job to the "#{@slave.name}" label}
end

When /^I enable concurrent builds$/ do
  step %{I check the "_.concurrentBuild" checkbox}
end

When /^I add a string parameter "([^"]*)"$/ do |name|
  parameter = @job.add_parameter "String Parameter"
  parameter.name = name
end

When /^I add a string parameter "([^"]*)" defaulting to "(.*?)"$/ do |name, default|
  parameter = @job.add_parameter "String Parameter"
  parameter.name = name
  parameter.default = default
end

When /^I disable the job$/ do
  @job.configure do
    @job.disable
  end
end

When /^I set artifacts? "([^"]*)" to archive$/ do |artifacts|
  @artifact_archiver = Jenkins::ArtifactArchiver.add(@job)
  @artifact_archiver.includes artifacts
end

When /^I set artifacts? "([^"]*)" to archive in the job configuration$/ do |artifacts|
  @job.configure do
    Jenkins::ArtifactArchiver.add(@job).includes artifacts
  end
end

When /^I set artifacts? "([^"]*)" to archive and exclude "([^"]*)" in the job configuration$/ do |include, exclude|
  @job.configure do
    archiver = Jenkins::ArtifactArchiver.add(@job)
    archiver.includes include
    archiver.excludes exclude
  end
end

When /^I want to keep only the latest successful artifacts$/ do
  @artifact_archiver.latest_only true
end

When /^I set (\d+) builds? to keep$/ do |number|
  step %{I check the "logrotate" checkbox}

  name = if @runner.jenkins_version < Gem::Version.new('1.503') then
    'logrotate_nums' else '_.numToKeepStr'
  end

  find(:xpath, "//input[@name='#{name}']").set(number)
end

When /^I schedule job to run periodically at "([^"]*)"$/ do |schedule|
  step 'I check the "hudson-triggers-TimerTrigger" checkbox'
  find(:path, '/hudson-triggers-TimerTrigger/spec').set(schedule)
end

When /^I use "([^"]*)" as custom workspace$/ do |workspace|
  @job.use_custom_workspace(workspace)
end

When /^I copy resource "([^"]*)" into workspace$/ do |resource|
  @job.copy_resource(resource, "")
end

When /^I copy resource "([^"]*)" into workspace as "([^"]*)"$/ do |resource, target|
  @job.copy_resource(resource, target)
end

File: src/main/java/org/jenkinsci/test/acceptance/steps/JobConfigurationSteps.java
package org.jenkinsci.test.acceptance.steps;

import cucumber.api.java.en.And;
import cucumber.api.java.en.When;
import hudson.util.VersionNumber;
import org.jenkinsci.test.acceptance.po.ArtifactArchiver;
import org.jenkinsci.test.acceptance.po.StringParameter;

import javax.inject.Inject;

import static org.jenkinsci.test.acceptance.cucumber.By2.*;

/**
 * @author Kohsuke Kawaguchi
 */
public class JobConfigurationSteps extends AbstractSteps {
    @Inject
    GeneralSteps generalsteps;

    @And("^I add a shell build step \"([^\"]*)\"$")
    public void I_add_a_shell_build_step(String shell) throws Exception {
        my.step = my.job.addShellStep(shell);
    }

    @And("^I enable concurrent builds$")
    public void I_enable_concurrent_builds() throws Exception {
        generalsteps.checkTheCheckbox("_.concurrentBuild");
    }

    @And("^I add a string parameter \"([^\"]*)\"$")
    public void I_add_a_string_parameter(String name) throws Exception {
        StringParameter p = my.job.addParameter(StringParameter.class);
        p.setName(name);
    }

    @When("^I disable the job$")
    public void I_disable_the_job() throws Throwable {
        my.job.configure();
        my.job.disable();
        my.job.save();
    }

    @And("^I set (\\d+) builds? to keep$")
    public void I_set_builds_to_keep(int n) throws Throwable {
        check("logrotate");
        String name = jenkins.getVersion().compareTo(new VersionNumber("1.503"))<0 ? "logrotate_nums" : "_.numToKeepStr";
        find(xpath("//input[@name='%s']",name)).sendKeys(String.valueOf(n));
    }

    @And("^I add always fail build step$")
    public void I_add_always_fail_build_step() throws Throwable {
        my.job.addShellStep("exit 1");
    }

    @When("^I add a shell build step \"([^\"]*)\" in the job configuration$")
    public void I_add_a_shell_build_step_in_the_job_configuration(String shell) throws Throwable {
        my.job.configure();
        my.job.addShellStep(shell);
        my.job.save();
    }

    @When("^I add a shell build step in the job configuration$")
    public void I_add_a_shell_build_step_in_the_job_configuration2(String shell) throws Throwable {
        I_add_a_shell_build_step_in_the_job_configuration(shell);
    }

    @And("^I set artifact \"([^\"]*)\" to archive$")
    public void I_set_artifact_to_archive(String artifacts) throws Throwable {
        my.job.addPublisher(ArtifactArchiver.class).includes(artifacts);
    }

    @And("^I set artifact \"([^\"]*)\" to archive in the job configuration$")
    public void I_set_artifact_to_archive_in_the_job_configuration(String includes) throws Throwable {
        my.job.configure();
        my.job.addPublisher(ArtifactArchiver.class).includes(includes);
        my.job.save();
    }

    @And("^I set artifact \"([^\"]*)\" to archive and exclude \"([^\"]*)\" in the job configuration$")
    public void I_set_artifact_to_archive_and_exclude_in_the_job_configuration(String include, String exclude) throws Throwable {
        my.job.configure();
        ArtifactArchiver aa = my.job.addPublisher(ArtifactArchiver.class);
        aa.includes(include).excludes(exclude);
        my.job.save();
    }
}
